### PR TITLE
feat: Add support for fine-tuning API

### LINKED ISSUE #57: Add support for Fine-tuning API
**Is your feature request related to a problem? Please describe.**
Mistral AI introduced the possibility to fine-tune their LLMs to even further increase their accuracy and performance. It was added to the JS client in [version 0.5.0](https://github.com/mistralai/client-js/releases/tag/0.5.0) and we should also support it to offer the full functionality of the Mistral AI API.

**Describe the solution you'd like**
Implement the "files" and "jobs" features based on the official JS client.

**Describe alternatives you've considered**
N/A

**Additional context**
Mistral docs - https://docs.mistral.ai/capabilities/finetuning/
Mistral JS client implementation - https://github.com/mistralai/client-js/commit/0d1cc9c9cf4b56feac057aecc0f9c1ec60d631dd

### PR DESCRIPTION
**Pull request by KacperKluka**
### Description

Added support for the fine-tuning API. This involved adding the `FilesClient` and `JobsClient` to make the SDK API as close to the official JS SDK as possible.

### Issue

Closes #57 

### Changes Made

- Added `FilesClient` that handles file related operations
- Added `JobsClient` that handles fine-tuning jobs related operations
- Changed network requests code to handle file uploads
- Added docs, examples and README entries about new features

### Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/nomtek/mistralai_client_dart/blob/main/CONTRIBUTING.md) document.
- [x] I have tested my changes.
- [x] I have added/updated documentation accordingly.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -148,6 +148,47 @@ if (toolCall != null && toolCall.type == 'function') {
...
+
+// List jobs
+final jobs = await client.jobs.list();
+
+// Retrieve a job
+final retrievedJob = await client.jobs.retrieve(jobId: createdJob.id);
+
+// Cancel a job
+final canceledJob = await client.jobs.cancel(jobId: createdJob.id);
+```
```
Comment (gawi151) [inline README.md]: I think it would be better to just link to file in `example` directory. Less things to maintain in the future ;)

File: lib/src/jobs_client.dart [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  /// Creates a new fine-tune job.
+  ///
+  /// [model] specifies the model to be fine-tuned.
+  /// The [trainingFiles] contains as a list of file IDs of previously
+  /// uploaded training files via [FilesClient.create]. An optional list of
+  /// [validationFiles] can be provided to improve the fine-tuning process.
+  /// The [hyperparameters] specify the fine-tuning parameters for the job.
+  /// The [suffix] is an optional suffix added to the fine-tuned model name.
+  /// The [integrations] is an optional list of integrations to enable for
+  /// the fine-tuning job.
```
Comment (gawi151) [inline lib/src/jobs_client.dart]: nit: a little bit better readability in the IDE

```suggestion
  /// Creates a new fine-tune job.
  ///
  /// [model] specifies the model to be fine-tuned.
  /// 
  /// The [trainingFiles] contains as a list of file IDs of previously
  /// uploaded training files via [FilesClient.create].
  /// 
  /// An optional list of [validationFiles] can be provided to improve
  /// the fine-tuning process.
  /// 
  /// The [hyperparameters] specify the fine-tuning parameters for the job.
  /// 
  /// The [suffix] is an optional suffix added to the fine-tuned model name.
  /// 
  /// The [integrations] is an optional list of integrations to enable for
  /// the fine-tuning job.
```

File: lib/src/models/jobs.dart [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+  @override
+  String toString() {
+    return 'TrainingParameters{trainingSteps: $trainingSteps, '
+        'learningRate: $learningRate}';
+  }
+}
+
+@immutable
+@JsonSerializable(fieldRename: FieldRename.snake)
```
Comment (gawi151) [inline lib/src/models/jobs.dart]: Same as previous about `includeIfNull: false`.

File: lib/src/models/jobs.dart [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  started,
+  running,
+  failed,
+  success,
+  cancelled,
+  cancellationRequested,
+}
+
+@immutable
+@JsonSerializable(fieldRename: FieldRename.snake)
```
Comment (gawi151) [inline lib/src/models/jobs.dart]: Could you please add includeIfNull: false to the JsonSerializable annotation? We encountered issues with a change in the Mistral API where fields with null values were being sent, and they were rejected because they should not have been included.

Just in case some new optional params will be added.

File: test/src/fakes.dart [hunk 5]
```diff
@@ -70,7 +70,8 @@ Future<Map<String, dynamic>> _getJsonRequestBody(
 ) async {
   final requestBytesStream = request.finalize();
   final requestBodyString = await requestBytesStream.bytesToString();
-  if (requestBodyString.isEmpty) {
+  // Multipart request body (e.g. when uploading a file) is not a JSON string
+  if (requestBodyString.isEmpty || !requestBodyString.startsWith('{')) {
```
Comment (gawi151) [inline test/src/fakes.dart]: Looks ok but what if it's json array? Probably we should also check for `[`. Or maybe there is some better way to detect json. Wdyt?

**Review by gawi151 — COMMENTED** [review]
Thx for the contribution ;) Great job. There are few bits and pieces where we can improve.

File: lib/src/models/jobs.dart (hunk 4 above)
Comment (KacperKluka) [reply lib/src/models/jobs.dart]: Sure 😁  Should I add it only in the 2 places you mentioned in the review or should they be added to each `@JsonSerializable` annotation? 🤔 https://github.com/nomtek/mistralai_client_dart/pull/59/commits/936d38f0d9b4884734df087435e17ff3b6614fd4

File: README.md (hunk 1 above)
Comment (KacperKluka) [reply README.md]: https://github.com/nomtek/mistralai_client_dart/pull/59/commits/406d31b3d5691a05b33854ed372cc20aad108a9b

File: lib/src/jobs_client.dart (hunk 2 above)
Comment (KacperKluka) [reply lib/src/jobs_client.dart]: https://github.com/nomtek/mistralai_client_dart/pull/59/commits/a69cce334c2a4bbbf8673bedfff5e6a59850fc45

File: test/src/fakes.dart (hunk 5 above)
Comment (KacperKluka) [reply test/src/fakes.dart]: Good point 👍 Since the multipart request body starts with the delimiter (a few `-` followed by a random string, e.g. "--2a8ae6ad-f4ad-4d9a-a92c-6d217011fe0f") I check if the body starts with `-` and if it does I assume it's not a JSON 😉 https://github.com/nomtek/mistralai_client_dart/pull/59/commits/33060093fc6f5b18a41e1218b7455bab70096f2c

File: lib/src/models/jobs.dart (hunk 4 above)
Comment (gawi151) [reply lib/src/models/jobs.dart]: I think those 2 places are enough for now. Basically we should have it whenever we are doing `toJson`.

**Review by gawi151 — APPROVED** [review]
LGTM ;)
